### PR TITLE
Item Pocket

### PR DESCRIPTION
**Pull request by MatusGuy**
The Item Pocket allows you to save a powerup for later use. If you collect a flower while having a bonus greater than `GROWUP_BONUS`, that flower gets equipped and the old flower gets stored in the top left corner of the screen. Now, the player can press the new `ITEM` control (usually Select/Back or Left Shift) to use the stored flower by throwing it up and catching it.

This pull request replaces the powerup stacking feature because this new solution is much more balanced. It also limits the amount of concurrent players to 4.

### CODE REVIEW

**Comment by tobbi** [comment]
Please fix initialization order:
```
/Users/runner/work/supertux/supertux/src/supertux/player_status.cpp:38:3: error: field 'bonus' will be initialized after field 'm_item_pockets' [-Werror,-Wreorder-ctor]
  bonus(num_players),
  ^~~~~~~~~~~~~~~~~~
  coins(START_COINS)
```

**Comment by tobbi** [comment]
Oh, right, this is a draft, so my comments were maybe a bit premature.

**Comment by tobbi** [comment]
```
/home/runner/work/supertux/supertux/src/badguy/boss.cpp:76:70: error: implicit conversion from 'int' to 'float' may lose precision [-Werror,-Wimplicit-int-float-conversion]
    float startpos = (context.get_width() - (m_hud_head->get_width() * m_max_lives)) / 2;
```

**Comment by swagtoy** [comment]
Has the ability to disable the itembox been discussed?

**Comment by bruhmoent** [comment]
> Has the ability to disable the itembox been discussed?

What do you mean by disabling it...? It's a core gameplay feature

**Comment by swagtoy** [comment]
> What do you mean by disabling it...? It's a core gameplay feature

Pardon? I meant that some level designers may want to disable it for themselves (or per world). I am curious on the consensus. I personally wouldnt enable it for my worlds

**Review by Frostwithasideofsalt — APPROVED** [review]
works, doesnt twerks  👩‍🔬

**Comment by Rusty-Box** [comment]
Found tiny bug: after I played through a level and went back to the worldmap, the item pocket visually disappears while on the worldmap. It does come back when entering a level, but goes away again afterwards.

File: src/control/game_controller_manager.cpp
Comment (Vankata453) [inline src/control/game_controller_manager.cpp:1]: You can remove the `sector.hpp` include.

File: src/control/joystick_manager.cpp
Comment (Vankata453) [inline src/control/joystick_manager.cpp:1]: You can remove the `sector.hpp` include.

File: src/supertux/player_status.cpp [hunk 1]
```diff
@@ -233,32 +257,87 @@ PlayerStatus::read(const ReaderMapping& mapping)
...
+  if (bonustype == NO_BONUS)
+    return;
+
+  m_item_pockets[player->get_id()] = NO_BONUS;
+
+  auto& powerup = Sector::get().add<PocketPowerUp>(bonustype, Vector(0,0));
+  Vector pos = player->get_pos();
+  pos.x = player->get_bbox().get_left();
+  pos.y = player->get_bbox().get_top() - powerup.get_bbox().get_height() - 15;
+  powerup.set_pos(pos);
```
Comment (Vankata453) [inline src/supertux/player_status.cpp:276]:
```suggestion
  powerup.set_pos(player->get_pos() - Vector(0.f, powerup.get_bbox().get_height() + 15));
```

File: src/supertux/player_status.cpp [hunk 2]
```diff
@@ -233,32 +257,87 @@ PlayerStatus::read(const ReaderMapping& mapping)
...
+      return true;
+    }
+  }
+
+  Level* level = Level::current();
+  int allowed = static_cast<Level::Setting>(level->m_allow_item_pocket);
+
+  if (allowed != Level::INHERIT)
+  {
+    return allowed == Level::ON ? true : false;
```
Comment (Vankata453) [inline src/supertux/player_status.cpp:313]:
```suggestion
    return allowed == Level::ON;
```

File: src/supertux/player_status.cpp [hunk 3]
```diff
@@ -233,32 +257,87 @@ PlayerStatus::read(const ReaderMapping& mapping)
...
+      return worldmap->is_item_pocket_allowed();
+    }
+    else
+    {
+      // This level is probably in a levelset, pick ON.
+      return true;
+    }
+  }
+
+  Level* level = Level::current();
```
Comment (Vankata453) [inline src/supertux/player_status.cpp:308]: Why not `GameSession::current()->get_current_level()`? `Level::current()` is funky IMO. Also you checked if `GameSession::current()` is valid beforehand.

File: src/supertux/player_status.cpp [hunk 4]
```diff
@@ -301,17 +377,55 @@ PlayerStatus::remove_player(int player_id)
...
-  max_air_time.resize(m_num_players, 0);
-  max_earth_time.resize(m_num_players, 0);
+  m_item_pockets.resize(m_num_players, NO_BONUS);
+}
+
+PlayerStatus::PocketPowerUp::PocketPowerUp(BonusType bonustype, Vector pos):
+  PowerUp(pos, PowerUp::get_type_from_bonustype(bonustype)),
+  m_cooldown_timer(),
+  m_blink_timer(),
+  m_visible(true)
```
Comment (Vankata453) [inline src/supertux/player_status.cpp:391]: Simply suggesting another newline here for some separation from the `PlayerStatus` class.

```suggestion


PlayerStatus::PocketPowerUp::PocketPowerUp(BonusType bonustype, Vector pos):
  PowerUp(pos, PowerUp::get_type_from_bonustype(bonustype)),
  m_cooldown_timer(),
  m_blink_timer(),
  m_visible(true)
```

File: src/supertux/player_status.hpp [hunk 5]
```diff
@@ -38,6 +42,7 @@ enum BonusType {
     example when switching maps in the worldmap) */
 class PlayerStatus final
 {
+
```
Comment (Vankata453) [inline src/supertux/player_status.hpp:45]:
```suggestion
```

File: src/supertux/player_status.hpp [hunk 6]
```diff
@@ -23,9 +23,13 @@
 #include <string>
 #include <vector>
 
+#include "object/powerup.hpp"
+#include "supertux/timer.hpp"
+
 class DrawingContext;
 class ReaderMapping;
 class Writer;
+class Player;
```
Comment (Vankata453) [inline src/supertux/player_status.hpp:32]:
```suggestion
class DrawingContext;
class Player;
class ReaderMapping;
class Writer;
```

File: src/supertux/player_status_hud.cpp [hunk 7]
```diff
@@ -16,12 +16,19 @@
...
 #include "supertux/player_status.hpp"
 #include "supertux/resources.hpp"
+#include "supertux/title_screen.hpp"
+#include "supertux/screen_manager.hpp"
 #include "video/drawing_context.hpp"
 #include "video/surface.hpp"
 #include "editor/editor.hpp"
+#include "worldmap/worldmap_sector.hpp"
+
+#include <iostream>
```
Comment (Vankata453) [inline src/supertux/player_status_hud.cpp:31]:
```suggestion

#include <iostream>

#include "sprite/sprite_manager.hpp"
#include "supertux/game_object.hpp"
#include "supertux/level.hpp"
#include "supertux/player_status.hpp"
#include "supertux/resources.hpp"
#include "supertux/title_screen.hpp"
#include "supertux/screen_manager.hpp"
#include "video/drawing_context.hpp"
#include "video/surface.hpp"
#include "editor/editor.hpp"
#include "worldmap/worldmap_sector.hpp"
```

File: src/supertux/sector.cpp [hunk 8]
```diff
@@ -589,6 +589,19 @@ Sector::is_free_of_specifically_movingstatics(float left, float top, float right
   return m_collision_system->is_free_of_specifically_movingstatics(Rectf(Vector(left, top), Vector(right, bottom)), nullptr);
 }
 
+void
+Sector::set_allow_item_pocket(bool allow)
+{
+  //FIXME
+  //m_level.m_allow_item_pocket = allow;
+}
```
Comment (Vankata453) [inline src/supertux/sector.cpp:597]: Hasn't been fixed.

File: src/supertux/sector.hpp [hunk 9]
```diff
@@ -176,6 +176,19 @@ class Sector final : public Base::Sector
...
+   * @description Allows/disallows usage of the Item Pocket.
+   * @param bool $allow
+   */
+  void set_allow_item_pocket(bool allow);
+
+  /**
+   * @scripting
+   * @description Checks if the Item Pocket is currently allowed.
+   */
+  bool is_item_pocket_allowed();
```
Comment (Vankata453) [inline src/supertux/sector.hpp:190]: Missing `const` qualifier.

**Review by Vankata453 — CHANGES_REQUESTED** [review]
Overall, amazing work on this PR!

**Encountered a bug:** The "Add Player" option in "Multiplayer Settings" does nothing.

Additionally, I believe that launching the powerup from the item pocket to the side (like a Tuxdoll), perhaps to the opposite side of where Tux is facing, would make utilizing the powerup from the item pocket require more concentration and consideration of where to utilize, since the player would also have to catch it, instead of waiting for it in the same spot.

I am also personally not a fan of the blinking effect that launched pocket powerups have. Maybe we could do without it?

**Comment by MatusGuy** [comment]
The blinking effect indicates that you can't pick it up yet. This is a mechanism that prevents immediately grabbing the object upon jumping.

Also, I think the general consensus is that it shouldn't move in x axis whatsoever but I might be wrong so I'll leave this up for discussion.

**Comment by MatusGuy** [comment]
Can you check if power up stacking was truly removed in case I missed something?

File: src/supertux/player_status.cpp (hunk 3 above)
Comment (MatusGuy) [reply src/supertux/player_status.cpp:308]: Is this value correct when in the editor?

Comment (Vankata453) [reply src/supertux/player_status.cpp]: Don't think so.

File: src/supertux/player_status.hpp (hunk 5 above)
Comment (tobbi) [reply src/supertux/player_status.hpp]: Peak change suggestion!

File: src/squirrel/supertux_api.cpp [hunk 10]
```diff
@@ -872,6 +904,9 @@ void register_supertux_scripting_api(ssq::VM& vm)
   level.addFunc("toggle_pause", &scripting::Level::toggle_pause);
   level.addFunc("pause_target_timer", &scripting::Level::pause_target_timer);
   level.addFunc("resume_target_timer", &scripting::Level::resume_target_timer);
+  level.addFunc("override_item_pocket", &scripting::Level::override_allow_item_pocket);
```
Comment (Vankata453) [inline src/squirrel/supertux_api.cpp:907]: The C++ function name is different. Make sure they both stay the same for accuracy in documentation.

```suggestion
  level.addFunc("override_item_pocket", &scripting::Level::override_item_pocket);
```

File: src/squirrel/supertux_api.cpp [hunk 11]
```diff
@@ -787,6 +787,38 @@ static void resume_target_timer()
   GameSession::current()->set_target_timer_paused(false);
 }
 
+/**
+ * @scripting
+ * @description Override the Item Pocket setting in the Level.
+ * @param string allow Can be "on", "off", or "inherit" (use the setting in the Level)
+ */
+static void override_allow_item_pocket(const std::string& allow)
```
Comment (Vankata453) [inline src/squirrel/supertux_api.cpp:795]: The C++ function name is different. Make sure they both stay the same for accuracy in documentation.

```suggestion
static void override_item_pocket(const std::string& allow)
```

**Review by Vankata453 — COMMENTED** [review]
Take a look at the comments below, as well as the following:

* multiplayer_player_menu.cpp:118 uses the same code as `GameSession::on_player_added` so it can be replaced with a call to that function.
* `GameSession::on_player_removed` shouldn't remove the player from `PlayerStatus`, since that will removed all the data for that player, when they may reconnect.
* ~~The "Add Player" bug from "Multiplayer Settings", as mentioned in the previous comment, is still present.~~

File: src/control/keyboard_config.cpp [hunk 12]
```diff
@@ -64,20 +66,6 @@ KeyboardConfig::KeyboardConfig() :
...
-  m_keymap[SDLK_s]         = {1, Control::DOWN};
-  m_keymap[SDLK_e]         = {1, Control::JUMP};
-  m_keymap[SDLK_q]         = {1, Control::ACTION};
-
-  m_keymap[SDLK_j]         = {2, Control::LEFT};
-  m_keymap[SDLK_l]         = {2, Control::RIGHT};
-  m_keymap[SDLK_i]         = {2, Control::UP};
-  m_keymap[SDLK_k]         = {2, Control::DOWN};
-  m_keymap[SDLK_o]         = {2, Control::JUMP};
-  m_keymap[SDLK_u]         = {2, Control::ACTION};
```
Comment (Vankata453) [inline src/control/keyboard_config.cpp:80]: Don't think there's any reason to remove the defaults for players 2 and 3. Perhaps put them back in and set a default control for the Item Pocket?

Comment (MatusGuy) [reply src/control/keyboard_config.cpp:80]: > Ok I think I remember why i did this
> Player 1 keybinds are literally incompatible with the rest - just picture it yourself
> You and your friends would be playing a game of twister on your keyboard

**Review by bruhmoent — APPROVED** [review]
Nice work - reviewed both the code and the in-game functionality

**Comment by Frostwithasideofsalt** [comment]
image
seems theres a bug with the item pocket that makes it so theres multiple item pockets when theres only one player

**Review by Frostwithasideofsalt — APPROVED** [review]
zero twerking found here, bucko! only working.

**Comment by Frostwithasideofsalt** [comment]
being that this seems to be done, could it be merged?

**Review by Vankata453 — APPROVED** [review]
Code-wise looks good now!

**Review by weluvgoatz — APPROVED** [review]
Works great!

**Pull request merged**